crates/project/src/git_store.rs

```diff
@@ -2643,36 +2647,56 @@ impl Repository {
             job_sender: Self::spawn_remote_git_worker(project_id, client, cx),
             askpass_delegates: Default::default(),
             latest_askpass_id: 0,
+            current_job: Default::default(),
         }
     }
 
     pub fn git_store(&self) -> Option<Entity<GitStore>> {
         self.git_store.upgrade()
     }
 
-    pub fn send_job<F, Fut, R>(&self, job: F) -> oneshot::Receiver<R>
+    pub fn send_job<F, Fut, R>(&self, status: Option<SharedString>, job: F) -> oneshot::Receiver<R>
     where
         F: FnOnce(RepositoryState, AsyncApp) -> Fut + 'static,
         Fut: Future<Output = R> + 'static,
         R: Send + 'static,
     {
-        self.send_keyed_job(None, job)
+        self.send_keyed_job(None, status, job)
     }
 
-    fn send_keyed_job<F, Fut, R>(&self, key: Option<GitJobKey>, job: F) -> oneshot::Receiver<R>
+    fn send_keyed_job<F, Fut, R>(
+        &self,
+        key: Option<GitJobKey>,
+        status: Option<SharedString>,
+        job: F,
+    ) -> oneshot::Receiver<R>
     where
         F: FnOnce(RepositoryState, AsyncApp) -> Fut + 'static,
         Fut: Future<Output = R> + 'static,
         R: Send + 'static,
     {
         let (result_tx, result_rx) = futures::channel::oneshot::channel();
+        let current_job = self.current_job.clone();
         self.job_sender
             .unbounded_send(GitJob {
                 key,
                 job: Box::new(|state, cx: &mut AsyncApp| {
                     let job = job(state, cx.clone());
                     cx.spawn(async move |_| {
+                        if let Some(s) = status.clone() {
+                            let mut current = current_job.lock();
+                            *current = Some((s.clone(), Instant::now()));
+                        }
                         let result = job.await;
+                        if let Some(s) = status {
+                            let mut current = current_job.lock();
+                            let old = current.take();
+                            debug_assert_eq!(
```

Comment:
@mikayla-maki this was the only part I wasn't 100% sure about so I left a debug assert in there.

If there's some way for the git worker to drop or cancel a job, or if it doesn't serialize the jobs, then you could end up in a situation where two jobs clobber each other when trying to set/clear `current_job`. afaik neither are possible today, but I assume it's fine to leave this in in-case a future change invalidates the invariant.

Reply:
I think I'd rather we structure this code so that this case is handled automatically. No need to depend on the serialization property for this feature.